src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/MemberTypeInfo.cs

```diff
@@ -142,63 +142,25 @@ internal TypeName GetArrayTypeName(ArrayInfo arrayInfo)
-                    PrimitiveType.Int32 => typeof(int),
-                    PrimitiveType.Int64 => typeof(long),
-                    PrimitiveType.SByte => typeof(sbyte),
-                    PrimitiveType.Single => typeof(float),
-                    PrimitiveType.TimeSpan => typeof(TimeSpan),
-                    PrimitiveType.DateTime => typeof(DateTime),
-                    PrimitiveType.UInt16 => typeof(ushort),
-                    PrimitiveType.UInt32 => typeof(uint),
-                    _ => typeof(ulong),
-                };
-
-                return primitiveType.BuildCoreLibArrayTypeName(arrayInfo.Rank);
-            case BinaryType.PrimitiveArray:
-                Type primitiveArrayType = ((PrimitiveType)additionalInfo!) switch
-                {
-                    PrimitiveType.Boolean => typeof(bool[]),
-                    PrimitiveType.Byte => typeof(byte[]),
-                    PrimitiveType.Char => typeof(char[]),
-                    PrimitiveType.Decimal => typeof(decimal[]),
-                    PrimitiveType.Double => typeof(double[]),
-                    PrimitiveType.Int16 => typeof(short[]),
-                    PrimitiveType.Int32 => typeof(int[]),
-                    PrimitiveType.Int64 => typeof(long[]),
-                    PrimitiveType.SByte => typeof(sbyte[]),
-                    PrimitiveType.Single => typeof(float[]),
-                    PrimitiveType.TimeSpan => typeof(TimeSpan[]),
-                    PrimitiveType.DateTime => typeof(DateTime[]),
-                    PrimitiveType.UInt16 => typeof(ushort[]),
-                    PrimitiveType.UInt32 => typeof(uint[]),
-                    _ => typeof(ulong[]),
-                };
-
-                return primitiveArrayType.BuildCoreLibArrayTypeName(arrayInfo.Rank);
-            case BinaryType.SystemClass:
-                return ((TypeName)additionalInfo!).BuildArrayTypeName(arrayInfo.Rank);
-            default:
-                Debug.Assert(binaryType is BinaryType.Class, "The parsers should reject other inputs");
-                return (((ClassTypeInfo)additionalInfo!).TypeName).BuildArrayTypeName(arrayInfo.Rank);
-        }
+            BinaryType.String => TypeNameHelpers.GetPrimitiveTypeName(PrimitiveType.String),
+            BinaryType.StringArray => TypeNameHelpers.GetPrimitiveSZArrayTypeName(PrimitiveType.String),
+            BinaryType.Primitive => TypeNameHelpers.GetPrimitiveTypeName((PrimitiveType)additionalInfo!),
+            BinaryType.PrimitiveArray => TypeNameHelpers.GetPrimitiveSZArrayTypeName((PrimitiveType)additionalInfo!),
+            BinaryType.Object => TypeNameHelpers.GetPrimitiveTypeName(TypeNameHelpers.ObjectPrimitiveType),
+            BinaryType.ObjectArray => TypeNameHelpers.GetPrimitiveSZArrayTypeName(TypeNameHelpers.ObjectPrimitiveType),
+            BinaryType.SystemClass => (TypeName)additionalInfo!,
+            _ => ((ClassTypeInfo)additionalInfo!).TypeName,
```

Comment:
We're losing an assert here. Is the assert no longer valid? Is there another consistency check we can add?

Reply:
We could re-add the assert here, but I don't see much value in that. The reader is well guarded against unknown types:
https://github.com/dotnet/runtime/blob/6e440de5b435aee32673376773c58652d1474f49/src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/Utils/BinaryReaderExtensions.cs#L32-L40

And if there is a type mismatch, the cast is going to throw anyway.